### PR TITLE
[core] gzip json parser

### PR DESCRIPTION
**Pull request by strcrzy**
to: @airbnb/streamalert-maintainers
size: small
issue: #16

# gzip-json parser
This StreamAlert parser parses deflate-compressed json-encoded strings, which is the format of the payload [provided](http://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/SubscriptionFilters.html#DestinationKinesisExample) by CloudWatch log subscriptions.

# nested records
the CloudWatch JSON payload contains nested records (the actual logs are at key `logEvents`) so this will only really be useful once 1784238 merges. i will wait until 1784238 has merged and rebase on top of that

### CODE REVIEW

**Comment by jacknagz** [comment]
This is great @strcrzy 😄 ! 

Did you experience issues with the KV/CSV parsers? Just trying to get some background on the added `try/except` blocks.

**Comment by strcrzy** [comment]
thanks! 

there's an issue where the `kv` and `csv` parsers were throwing exceptions when they attempted to parse the compressed payloads (they expect string-like payloads and choke on the binary data). 

this will happen when a source declares both (`kv` or `csv`) and `gzip-json` log types, in two situations that i am aware of:
1. `kv`/`csv` log types are declared before `gzip-json` one; these parsers are then attempted first
2. `gzip-json` log type is declared first, but the payload doesn't match the schema; the `kv`/`csv` parsers are then attempted next

in each of these situations it seems appropriate to catch the exception and return `False`

**Comment by jacknagz** [comment]
Awesome, thanks for the background.  Once the nested record parsing support merges, could you also add some tests for this new parser?

**Comment by strcrzy** [comment]
:+1: definitely

File: stream_alert/parsers.py [hunk 1]
```diff
@@ -134,6 +143,22 @@ def _parse_records(self, json_payload):
                         json_records.append(record)
                 elif isinstance(val, dict):
                     return self._parse_records(json_payload[key])
+        # if hints are present, use them to find the nested records
+        elif (hints and len(hints)):
+            records_jsonpath = jsonpath_rw.parse(hints['records'])
+            envelope_schema = hints.get('envelope', {})
+            if len(envelope_schema):
+                self.schema.update({u"envelope": envelope_schema})
```
Comment (jacknagz) [inline stream_alert/parsers.py]: why make the key `unicode` vs `str`?

File: stream_alert/classifier.py [hunk 2]
```diff
@@ -220,7 +220,6 @@ def _parse(self, payload, data):
         Returns:
             A boolean representing the success of the parse.
         """
-        logger.debug(data)
```
Comment (strcrzy) [inline stream_alert/classifier.py]: when this line attempts to log a binary payload, the lambda goes 💥

Comment (jacknagz) [inline stream_alert/classifier.py]: interesting, ok!

File: stream_alert/parsers.py [hunk 3]
```diff
@@ -114,7 +119,9 @@ def _key_check(self, json_records):
                 json_records.remove(json_record)
 
     def _parse_records(self, json_payload):
-        """Iterate over a json_payload to determine if nested records exist
+        """Iterate over a json_payload.
```
Comment (strcrzy) [inline stream_alert/parsers.py]: 👍

File: stream_alert/parsers.py (hunk 1 above)
Comment (strcrzy) [inline stream_alert/parsers.py]: i'm not sure why i put this here. removing it.

File: stream_alert/parsers.py [hunk 4]
```diff
@@ -124,7 +131,9 @@ def _parse_records(self, json_payload):
             A list of dict JSON payloads
         """
         json_records = []
-
+        hints = self.options.get('hints',{})
```
Comment (strcrzy) [inline stream_alert/parsers.py]: :+1:

**Comment by strcrzy** [comment]
added some docs!

File: docs/source/conf-schemas.rst [hunk 5]
```diff
@@ -146,6 +146,41 @@ This promotes Rule safety, but requires additional time to define the schemas
 
 .. warning:: In Option 2, the schema definition is flexible, but Rule safety is lost because you'll need to use defensive programming when accessing and analyzing fields in `columns`. The use of `req_subkeys` will be advised, see Rules for more details
 
+
+JSON Example: VPC Flow Logs
+---------------------------
+
+VPC flow logs can be delivered via CloudWatch to a kinesis stream. Though they are compressed with deflate, we can use the special ``gzip-json`` parser to decompress them before parsing.CloudWatch logs are delivered as a nested record, so we will need to pass ``hints`` to JSON parser to properly find the nested records::
```
Comment (jacknagz) [inline docs/source/conf-schemas.rst]: `...them before parsing.Cloudwatch`

pls fix spacing after `parsing.`

**Comment by strcrzy** [comment]
stripped previous nesting implementation, updated docs to reflect the JSONPath approach

File: stream_alert/parsers.py [hunk 6]
```diff
@@ -124,16 +135,23 @@ def _parse_records(self, json_payload):
...
+                self.schema.update({"envelope": envelope_schema})
+                envelope_keys = envelope_schema.keys()
+                envelope_jsonpath = jsonpath_rw.parse("$." + ",".join(envelope_keys))
+                envelope_matches = [match.value for match in envelope_jsonpath.find(json_payload)]
+                envelope = dict(zip(envelope_keys, envelope_matches))
+
+            for match in records_jsonpath.find(json_payload):
+                record = match.value
+                if len(envelope):
+                    record.update({"envelope":envelope})
```
Comment (jacknagz) [inline stream_alert/parsers.py]: space after "envelope"

**Pull request merged**